cmd/gwi-api/main.go

```diff
@@ -0,0 +1,350 @@
+	listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", port))
+	if err != nil {
+		return errors.Wrapf(err, "failed to listen on port %d", port)
+	}
+
+	e := echo.New()
+	e.GET("/", handler)
+
+	logger.Info("healthcheck server starting...", zap.Int("port", port))
+
+	go http.Serve(listener, e)
+	return nil
+}
+
+func runAPIServer(logger *zap.Logger, config *Config, a *api.API) error {
+	// Init Echo.
+	e := initEcho(logger, config, a)
+
+	// Init the HTTP server.
+	server := initHTTPServer(config, e)
+
+	// Keep processing requests until interrupted.
+	logger.Info("API server starting...",
+		zap.String("host", config.HTTPHost), zap.Int("port", config.HTTPPort))
+
+	return runServerUntilInterrupted(logger, server)
+}
+
+func initHTTPServer(config *Config, handler http.Handler) *http.Server {
+	return &http.Server{
+		Addr:              fmt.Sprintf("%v:%v", config.HTTPHost, config.HTTPPort),
+		Handler:           handler,
+		ReadTimeout:       config.HTTPReadTimeout,
+		ReadHeaderTimeout: config.HTTPReadHeaderTimeout,
+		WriteTimeout:      config.HTTPWriteTimeout,
+		IdleTimeout:       config.HTTPIdleTimeout,
+		MaxHeaderBytes:    config.HTTPMaxHeaderBytes,
+	}
+}
+
+func initEcho(logger *zap.Logger, config *Config, a *api.API) *echo.Echo {
+	e := echo.New()
+	e.Pre(middleware.RemoveTrailingSlash())
+	e.Use(middleware.RequestID())
+	e.Use(middleware.Recover())
+
+	// Logging kinda sucks with Echo.
+	// Not feeling like implementing echo.Logger interface,
+	// it is not really compatible with zap.Logger approach.
+	if config.DebugEnabled {
+		// Log every request when debug is enabled.
+		e.Use(func(next echo.HandlerFunc) echo.HandlerFunc {
+			return func(c echo.Context) error {
+				req := c.Request()
+				res := c.Response()
+				reqID := res.Header().Get(echo.HeaderXRequestID)
+				logger.Debug(
+					"request received",
+					zap.String("request_id", reqID),
+					zap.Object("request", types.HTTPRequest{R: req}),
+				)
+
+				start := time.Now()
+				err := next(c)
+				delta := time.Since(start)
+
+				code := res.Status
+				if err != nil {
+					if he, ok := err.(*echo.HTTPError); ok {
+						code = he.Code
+					} else {
+						code = http.StatusInternalServerError
+					}
+				}
+
+				logger.Debug(
+					"response sent",
+					zap.String("request_id", reqID),
+					zap.Int("status_code", code),
+					zap.Duration("duration", delta),
+					zap.Error(err),
+				)
+				return err
+			}
+		})
+	} else {
+		// Otherwise just log errors.
+		e.Use(func(next echo.HandlerFunc) echo.HandlerFunc {
+			return func(c echo.Context) error {
+				err := next(c)
+				if err != nil {
+					req := c.Request()
+					res := c.Response()
+					logger.Error(
+						"request failed",
+						zap.Error(err),
+						zap.String("request_id", res.Header().Get(echo.HeaderXRequestID)),
+						zap.Object("request", types.HTTPRequest{R: req}),
+					)
+				}
+				return err
+			}
+		})
+	}
+
+	jwtMiddleware := middleware.JWT([]byte(config.JWTSecret))
```

Comment:
Can you explain what this middleware does?

Reply:
It ensures that there is a valid JWT token present in the request. In case there isn't, it responds with `Unauthorized`.

Reply:
And in case you want to return some custom body or http status, does it allow you?

Reply:
You can set `ErrorHandler` on https://godoc.org/github.com/labstack/echo/middleware#JWTConfig and use `middleware.JWTWithConfig` middleware factory.